### PR TITLE
Refactor tools.cpp to homogenize system tools policies

### PR DESCRIPTION
**Pull request by ras0219-msft**
This PR subsumes #181.

This PR refactors the internal implementation of `ToolsCache` to achieve several goals:
1. Makes parsing of vcpkgTools.xml testable
2. Remove `VcpkgPaths` from the `ToolsCache` / `ToolsProvider` interface by injecting the specific data required
3. Downgrade `VcpkgPaths` to ToolsCache in the `archives.cpp` functions
4. Localize many existing and all new messages in `tools.cpp`
5. Centralize handling of `VCPKG_FORCE_SYSTEM_BINARIES` & `VCPKG_FORCE_DOWNLOADED_BINARIES`

/* todo: add more examples of error messages */

### CODE REVIEW

File: include/vcpkg/fwd/vcpkgpaths.h [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 namespace vcpkg
 {
+    struct DownloadManager;
```
Comment (BillyONeal) [inline include/vcpkg/fwd/vcpkgpaths.h:5]: Don't think this is the right fwd header for that?

File: src/vcpkg/tools.cpp [hunk 2]
```diff
@@ -726,171 +598,291 @@ coscli version v0.11.0-beta
...
+
+                return PathAndVersion{candidate, *version};
+            }
 
-                    return std::move(tars[0]);
-                }
+            return nullopt;
+        }
+
+        Path download_tool(StringView tool_name, const ToolData& tool_data) const
```
Comment (BillyONeal) [inline src/vcpkg/tools.cpp:648]: It seems like tool_name should be part of tool_data :)

No change requested.

File: src/vcpkg/export.ifw.cpp [hunk 3]
```diff
@@ -354,7 +354,8 @@ namespace vcpkg::Export::IFW
 
         void do_repository(const std::string& export_id, const Options& ifw_options, const VcpkgPaths& paths)
         {
-            const Path& repogen_exe = paths.get_tool_exe(Tools::IFW_REPOGEN);
+            Path repogen_exe = paths.get_tool_exe(Tools::IFW_INSTALLER_BASE);
+            repogen_exe.replace_filename("repogen.exe");
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/export.ifw.cpp:358]: Good point, I didn't consider that this changes `vcpkg fetch ifw_repogen`. However, I think this is still 100% worth removing the special case from `get_tool_exe()`. In the future we may want to consider unpacking tools based on their archive SHAs, which would make this kind of "reuse" very straightforward: just redeclare the same archive with a different exe path.

File: src/vcpkg/tools.cpp [hunk 4]
```diff
@@ -726,171 +598,291 @@ coscli version v0.11.0-beta
 
     struct ToolCacheImpl final : ToolCache
     {
-        RequireExactVersions abiToolVersionHandling;
-        vcpkg::Cache<std::string, Path> system_cache;
-        vcpkg::Cache<std::string, Path> path_only_cache;
+        Filesystem& fs;
+        const DownloadManager& downloader;
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/tools.cpp:602]: The current structure of these individual manager objects is that, for most operations, VcpkgPaths is the direct owner of all of them and injects cross references as required. It ensures that they all share ~the same lifetime (that of the VcpkgPaths object). I also want DownloadManager to be morally an interface for all consumers; they shouldn't need to see the config object.

Would moving these individual managers to `shared_ptr<>`s be preferable for you? I personally am not a big fan of using `shared_ptr<>`s here since I think they indicate much more ambiguity in ownership than intended. Proposal 2: what if it actually stored a `weak_ptr<>`? That would downgrade use-after-free to a crash, but still preserve the more "strict" ownership model I'd like to pursue.

Finally, I don't like passing the filesystem in these interfaces since it is not an _essential_ aspect of the call. The current interface can perfectly be implemented via two simple `map`s with no filesystem operations at all. Contrast that to the `DownloadManager`, which has the contract that it will _create_ the downloaded file (or read the uploaded file) and so it is essential to the function contract that there is a filesystem.

So to summarize counter proposals in order of my preference:
1. as-is References. Optionally add some debug-mode-only machinery that helps to catch use-after-free issues like zeroing out the members.
2. Store `weak_ptr<>`s to one or both of Filesystem and DownloadManager. This change is only to catch use-after-free and is not intended to indicate ownership.
3. Store `shared_ptr<>`s to one or both of them.
4. Concoct some custom type that acts like `weak_ptr` but only lets you get a pointer out, not a full shared_ptr. This ensures unique ownership in the typesystem while still catching use-after-free (modulo threads, but we have bigger problems if these objects have data races on them).

File: src/vcpkg/tools.cpp [hunk 5]
```diff
@@ -726,171 +598,291 @@ coscli version v0.11.0-beta
...
-                        return {"ninja", "0"};
-                    }
-                    return get_path(paths, NinjaProvider());
-                }
-                if (tool == Tools::POWERSHELL_CORE)
+            return xml_config_contents.get_lazy(
+                [this]() { return this->fs.read_contents(this->xml_config, VCPKG_LINE_INFO); });
+        }
+
+        virtual const Path& get_tool_path(StringView tool) const override { return get_tool_pathversion(tool).p; }
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/tools.cpp:714]: I like this overload because 90% of callers just want the path so they can proceed to build a command out of it.

If you would like to see `get_tool_version()` replaced with `get_tool_pathversion().version` while keeping `get_tool_path()`, I can do that.

File: src/vcpkg/tools.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/tools.cpp:602]: I would rank those:

```
3
1
2







4
```

[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/tools.cpp:602]: (And if 3 ends up being obnoxious work, this isn't an over my dead body thing: as I mentioned in chat I think the right thing in the longer term we need change the design to avoid circularity)

File: src/vcpkg/tools.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/tools.cpp:714]: You could make it nonvirtual and in the interface if it's to make callers prettier.

File: src/vcpkg/tools.cpp [hunk 6]
```diff
@@ -609,13 +625,13 @@ coscli version v0.11.0-beta
         vcpkg::Cache<std::string, PathAndVersion> path_version_cache;
 
         ToolCacheImpl(Filesystem& fs,
-                      const DownloadManager& downloader,
+                      std::shared_ptr<const DownloadManager>&& downloader,
```
Comment (BillyONeal) [inline src/vcpkg/tools.cpp:628]:
```suggestion
                      const std::shared_ptr<const DownloadManager>& downloader,
```

No reason to block non-temporaries

**Review by BillyONeal — APPROVED** [review]

File: src/vcpkg/tools.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/tools.cpp:730]: That would be fine by me -- I thought you hated having _any_ code in the interface though?
